cfdm/functions.py

```diff
@@ -998,38 +747,86 @@ def copy(self):
         return out
 
 
-class Configuration(dict):
+class Configuration(dict, metaclass=DocstringRewriteMeta):
     '''A dictionary-like container for the global constants with context
     manager support.
 
-    Initialization is as for a `dict`, and all of the `dict` methods
-    are available with the same behaviours (`clear`, `copy`,
+    Initialization is as for a `dict`, and nearly all of the `dict`
+    methods are available with the same behaviours (`clear`,
     `fromkeys`, `get`, `items`, `keys`, `pop`, `popitem`,
     `setdefault`, `update`, `values`):
 
-       >>> c = cfdm.Configuration(atol=0.1, rtol=0.2, log_level='INFO')
+       >>> c = {{package}}.{{class}}(atol=0.1, rtol=0.2, log_level='INFO')
        >>> c
-       <Configuration: {'atol': 0.1, 'rtol': 0.2, 'log_level': 'INFO'}>
+       <{{repr}}{{class}}: {'atol': 0.1, 'rtol': 0.2, 'log_level': 'INFO'}>
        >>> print(c)
        {'atol': 0.1, 'rtol': 0.2, 'log_level': 'INFO'}
        >>> c.pop('atol')
        0.1
        >>> c
-       <Configuration: {'rtol': 0.2, 'log_level': 'INFO'}>
+       <{{repr}}{{class}}: {'rtol': 0.2, 'log_level': 'INFO'}>
        >>> c.clear()
        >>> c
-       <Configuration: {}>
+       <{{repr}}{{class}}: {}>
+
+    The `copy` method return a deep copy, rather than a shallow one.
+
+    **Context manager**
+
+    The `{{class}}` instance can be used as a context manager that
+    upon exit executes the function defined by the `_func` attribute,
+    with the class itself as input *kwargs* parameters. For example,
+    the `{{class}}` instance ``c`` would execute ``c._func(**c)`` upon
+    exit.
 
     .. versionadded:: (cfdm) 1.8.8.0
 
     '''
-    __slots__ = ('_func',)
-
     def __new__(cls, *args, **kwargs):
+        '''Must override this method in subclasses.
+
+        '''
         instance = super().__new__(cls)
         instance._func = configuration
         return instance
 
+    def __docstring_substitutions__(self):
```

Comment:
When checking the docs built off this branch, I've noticed any dunderscore methods aren't currently processed and included in the API reference. If we want to include them, we'll need to tell Sphinx we want them. This worked for me (I'm using `v3.2.1`):

```diff
diff --git a/docs/source/conf.py b/docs/source/conf.py
index 041303e43..fba9321bb 100755
--- a/docs/source/conf.py
+++ b/docs/source/conf.py
@@ -130,6 +130,7 @@ autodoc_member_order = 'groupwise'
 autodoc_default_options = {
     'members': True,
     'inherited-members': True,
+    'special-members': True,
     'show-inheritance': True,
 }
```

Reply:
I'll need to think about this one, just because I can't remember what the implications are for the appearance of classes in general ....

Reply:
Sounds wise :) Please take your time deciding, after all it is easy to add this or similar in at a later date.

Reply:
Hi Sadie, when I set `special-members` at v2.3.1 nothing happens, possibly because we set `no-members` on all of our `autolcass` directives? I tend to only document the dunderscore methods that have been overridden, but that's not  consistently applied, either. I'll try at 3.2.1 and see what happens ...

(pause of a few minutes)

... Wow. 3.2.1 destroyed the docs, with or without `special-members` in different ways. It's actually trashed all of the files  in `docs/source/class`! (not problem in getting them back, though) 

Can we raise this as a separate issue, that also will address the "right" version of sphinx to use? I'm sure it should be newer than my 2.3.1, and I may just need to update something else in my environment (I hope)